language/move-lang/src/naming/ast.rs

```diff
@@ -376,34 +384,55 @@ impl BuiltinFunction_ {
 }
 
 impl BaseType_ {
+    pub fn kind_(&self) -> Kind_ {
+        match self {
+            Self::Param(_) | Self::Var(_) | Self::Anything => Kind_::Unknown,
+            Self::Apply(kind_opt, _, _) => match kind_opt {
+                Some(sp!(_, k)) => k.clone(),
+                None => Kind_::Unknown,
+            },
+        }
+    }
+
     pub fn anything(loc: Loc) -> BaseType {
         sp(loc, BaseType_::Anything)
     }
 
-    pub fn builtin(loc: Loc, b_: BuiltinTypeName_) -> BaseType {
-        let kind = sp(loc, b_.kind());
+    pub fn builtin(loc: Loc, b_: BuiltinTypeName_, ty_args: Vec<BaseType>) -> BaseType {
+        use BuiltinTypeName_::*;
+
+        let kind = match b_ {
+            U8 | U64 | U128 | Address | Bool | Bytearray => Kind_::Unrestricted,
+            // TODO: this can panic.
+            Vector => ty_args[0].value.kind_(),
```

Comment:
This is your bug. This needs to be `None`

Due to type inference, you cannot inspect the kind at this point. (in any ast past typing, it is fine to do so)